ocaml-lsp-server/src/document_store.ml

```diff
@@ -1,36 +1,154 @@
 open! Import
 open Fiber.O
 
-type t = (Uri.t, Document.t) Table.t
+type server = Server : 'a Server.t Fdecl.t -> server
 
-let make () = Table.create (module Uri) 50
+type doc =
+  { (* invariant: if [document <> None], then no promotions are active *)
```

Comment:
1. what does it mean for promotions to be "active"? 

2. What does it mean for a document to have more than one promotion? Can those be like one promotion from formatting, some other from expect tests, etc? 

2.1. If yes, what do we expect `Promote` code action to do? Which promotion will be applied? 

(I hope questions make sense)

Reply:
1. There's a promotion being offered for this uri by some instance of dune
2. It is possible if you're running more than one instance of dune. E.g. you're building a project separately, and as a part of a bigger workspace.
2.1. There would be 2 promotion actions offered. They would look identical, so it would be hard to choose one. We should fix that of course.